### ISSUE TITLE
Rustdoc doesn't warn if links to associated items are ambiguous if primitives are involved

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust#83862]
Rustdoc should give an error if the base res is ambiguous, but it doesn't:
```
$ cat primitive-ambiguous.rs 
//! [usize::MAX]

mod usize {
  const MAX: usize = 1;
}
$ rustdoc primitive-ambiguous.rs
$
```

This was originally discovered through the special case
```rust
//! [usize::MAX]

#[doc(primitive = "usize")]
mod usize {}
```
which I think *should* continue working, rustdoc shouldn't give an ambiguity error if the module has `doc(primitive)` on it.

_Originally posted by @jyn514 in https://github.com/rust-lang/rust/pull/83849#discussion_r606850759_

## Meta

HEAD is 2616ab1c57e2d69f989307389b27ee996ed82575

**Comment by jyn514** [comment]
Oof I just realized this is a lot harder than I expected. For the case where `usize` is a module, `usize::MAX` resolves normally with `resolve_path`, before the hacky string splitting for `resolve_associated_item`. So to support `prim@usize::MAX` I need to do the string splitting *before* resolve_path ever runs, and call `resolve_primitive_associated_item` instead.

@Manishearth btw do you *want* that `prim@usize::MAX` syntax to work? Or should it just not be possible to disambiguate the two? It does seem weird for the disambiguator to refer to the root res `usize` instead of `MAX`, that's inconsistent with how it works eveywhere else.

**Comment by Manishearth** [comment]
> @Manishearth btw do you _want_ that `prim@usize::MAX` syntax to work? Or should it just not be possible to disambiguate the two? It does seem weird for the disambiguator to refer to the root res `usize` instead of `MAX`, that's inconsistent with how it works eveywhere else.

Hmm, this sounds like a pretty big complication; unsure if we should. 

We can just special case the primitives and do both lookups.

**Comment by jyn514** [comment]
> We can just special case the primitives and do both lookups.

Sorry, I don't follow - what should happen if it's ambiguous (both resolve)?

**Comment by jyn514** [comment]
I would also be ok with saying "don't name modules after primitives", that seems confusing anyway.

**Comment by Manishearth** [comment]
I mean, don't name modules after primitives such that the names clash. Fine with wontfixing
